hack/deploy-karmada.sh

```diff
@@ -118,20 +121,32 @@ kubectl apply -f "${REPO_ROOT}/artifacts/deploy/karmada-etcd.yaml"
 # Wait for karmada-etcd to come up before launching the rest of the components.
 util::wait_pod_ready "${ETCD_POD_LABEL}" "${KARMADA_SYSTEM_NAMESPACE}"
 
+
+# If it provided a karmada API Server IP we can access karmada API Server (cluster by kind), we will create a ClusterIP Service
+# Or we need to create a LoadBalancer service so that we can access karmada API Server outside the karmada-host cluster
+KARMADA_APISERVER_SERVICE_TYPE="ClusterIP"
+if [[ -z "${KARMADA_APISERVER_IP}" ]]; then
+  KARMADA_APISERVER_SERVICE_TYPE="LoadBalancer"
+fi
+
 # deploy karmada apiserver
-kubectl apply -f "${REPO_ROOT}/artifacts/deploy/karmada-apiserver.yaml"
+TEMP_PATH_APISERVER=$(mktemp -d)
+cp "${REPO_ROOT}"/artifacts/deploy/karmada-apiserver.yaml "${TEMP_PATH_APISERVER}"/karmada-apiserver.yaml
+sed -i "s/{{service_type}}/${KARMADA_APISERVER_SERVICE_TYPE}/g" "${TEMP_PATH_APISERVER}"/karmada-apiserver.yaml
+echo -e "Apply dynamic rendered apiserver service in ${TEMP_PATH_APISERVER}/karmada-apiserver.yaml.\n"
+kubectl apply -f "${TEMP_PATH_APISERVER}"/karmada-apiserver.yaml
 
 # Wait for karmada-apiserver to come up before launching the rest of the components.
 util::wait_pod_ready "${APISERVER_POD_LABEL}" "${KARMADA_SYSTEM_NAMESPACE}"
 
-if [[ -z "${3-}" ]]; then
-  KARMADA_APISERVER_IP=$(kubectl get service karmada-apiserver -n karmada-system -o jsonpath='{.spec.clusterIP}')
-else
-  KARMADA_APISERVER_IP=$3
+if [[ -z "${KARMADA_APISERVER_IP}" ]]; then
+  util::wait_service_external_ip "karmada-apiserver" "${KARMADA_SYSTEM_NAMESPACE}"
+  KARMADA_APISERVER_IP=$SERVICE_EXTERNAL_IP
+  sleep 3 # normally the load balancer ip need a few seconds to warm up
```

Comment:
Hi, @lfbear I don't quite understand why we need to sleep here since we have waited the external ip of the service ready in the wait_service_external_ip function.

Reply:
The function wait_service_external_ip should wait for the external IP ready originally, but I found in some public cloud it can't ping or curl while the external IP just came out, so I add 3 seconds additionally for that

Reply:
> The function wait_service_external_ip should wait for the external IP ready originally, but I found in some public cloud it can't ping or curl while the external IP just came out, so I add 3 seconds additionally for that

How about to add a check in the wait_service_external_ip function? Such as wait until the external IP can ping or curl. Maybe in some in some public cloud, 3 seconds are not enough.

Reply:
It's a good idea, I will add it in the wait_service_external_ip function